### ISSUE TITLE
How to use the MovieLens datasets in original split(5-folds) without valid data

### ISSUE THREAD

**Issue by congx961** [RUCAIBox/RecBole#586]
Hello,
I wanna use use the MovieLens datasets in original split(5-folds) without valid data, but I've noticed that the `spilt_ratio` in [evaluation setting](https://recbole.io/docs/user_guide/config_settings.html#basic-parameters) would always take affect even though I set it as [0.8, 0.2, 0](which actually convert into [0.8, 0.1, 0.1]). 

How can I make it?

**Comment by tsotfsk** [comment]
Hi! @congx961  Thanks for your question. At present, it is not easy to set two-phase splitting by configuring some parameters. So, the following code may help you and we will support this feature in the future versions.



1. in `\recbole\quick_start\quick_start.py`, line **41~55**

```python
# dataset splitting
train_data, valid_data = data_preparation(config, dataset)

# model loading and initialization
model = get_model(config['model'])(config, train_data).to(config['device'])
logger.info(model)

# trainer loading and initialization
trainer = get_trainer(config['MODEL_TYPE'], config['model'])(config, model)

# model training
best_valid_score, best_valid_result = trainer.fit(train_data, valid_data, saved=saved)

# model evaluation
test_result = trainer.evaluate(valid_data, load_best_model=saved)
```

2. in `\recbole\data\utils.py`, you can repalce the function `data_preparation` as belows:

```python
def data_preparation(config, dataset, save=False):
    """Split the dataset by :attr:`config['eval_setting']` and call :func:`dataloader_construct` to create
    corresponding dataloader.

    Args:
        config (Config): An instance object of Config, used to record parameter information.
        dataset (Dataset): An instance object of Dataset, which contains all interaction records.
        save (bool, optional): If ``True``, it will call :func:`save_datasets` to save split dataset.
            Defaults to ``False``.

    Returns:
        tuple:
            - train_data (AbstractDataLoader): The dataloader for training.
            - valid_data (AbstractDataLoader): The dataloader for validation.
            - test_data (AbstractDataLoader): The dataloader for testing.
    """
    model_type = config['MODEL_TYPE']

    es_str = [_.strip() for _ in config['eval_setting'].split(',')]
    es = EvalSetting(config)

    kwargs = {}
    if 'RS' in es_str[0]:
        kwargs['ratios'] = config['split_ratio']
        if kwargs['ratios'] is None:
            raise ValueError('`ratios` should be set if `RS` is set')
    if 'LS' in es_str[0]:
        kwargs['leave_one_num'] = config['leave_one_num']
        if kwargs['leave_one_num'] is None:
            raise ValueError('`leave_one_num` should be set if `LS` is set')
    kwargs['group_by_user'] = config['group_by_user']
    getattr(es, es_str[0])(**kwargs)

    if es.split_args['strategy'] != 'loo' and model_type == ModelType.SEQUENTIAL:
        raise ValueError('Sequential models require "loo" split strategy.')

    builded_datasets = dataset.build(es)
    train_dataset, valid_dataset = builded_datasets
    phases = ['train', 'test']

    if save:
        save_datasets(config['checkpoint_dir'], name=phases, dataset=builded_datasets)

    kwargs = {}
    if config['training_neg_sample_num']:
        es.neg_sample_by(config['training_neg_sample_num'])
        if model_type != ModelType.SEQUENTIAL:
            sampler = Sampler(phases, builded_datasets, es.neg_sample_args['distribution'])
        else:
            sampler = RepeatableSampler(phases, dataset, es.neg_sample_args['distribution'])
        kwargs['sampler'] = sampler.set_phase('train')
        kwargs['neg_sample_args'] = copy.deepcopy(es.neg_sample_args)
        if model_type == ModelType.KNOWLEDGE:
            kg_sampler = KGSampler(dataset, es.neg_sample_args['distribution'])
            kwargs['kg_sampler'] = kg_sampler
    train_data = dataloader_construct(
        name='train',
        config=config,
        eval_setting=es,
        dataset=train_dataset,
        dl_format=config['MODEL_INPUT_TYPE'],
        batch_size=config['train_batch_size'],
        shuffle=True,
        **kwargs
    )

    kwargs = {}
    if len(es_str) > 1 and getattr(es, es_str[1], None):
        getattr(es, es_str[1])()
        if 'sampler' not in locals():
            sampler = Sampler(phases, builded_datasets, es.neg_sample_args['distribution'])
        kwargs['sampler'] = [sampler.set_phase('test')]
        kwargs['neg_sample_args'] = copy.deepcopy(es.neg_sample_args)
    valid_data = dataloader_construct(
        name='evaluation',
        config=config,
        eval_setting=es,
        dataset=valid_dataset,
        batch_size=config['eval_batch_size'],
        **kwargs
    )

    return train_data, valid_data
```

**Comment by congx961** [comment]
> Hi! @congx961 Thanks for your question. At present, it is not easy to set two-phase splitting by configuring some parameters. So, the following code may help you and we will support this feature in the future versions.
> 
> 1. in `\recbole\quick_start\quick_start.py`, line **41~55**
> 
> ```python
> # dataset splitting
> train_data, valid_data = data_preparation(config, dataset)
> 
> # model loading and initialization
> model = get_model(config['model'])(config, train_data).to(config['device'])
> logger.info(model)
> 
> # trainer loading and initialization
> trainer = get_trainer(config['MODEL_TYPE'], config['model'])(config, model)
> 
> # model training
> best_valid_score, best_valid_result = trainer.fit(train_data, valid_data, saved=saved)
> 
> # model evaluation
> test_result = trainer.evaluate(valid_data, load_best_model=saved)
> ```
> 
> 1. in `\recbole\data\utils.py`, you can repalce the function `data_preparation` as belows:
> 
> ```python
> def data_preparation(config, dataset, save=False):
>     """Split the dataset by :attr:`config['eval_setting']` and call :func:`dataloader_construct` to create
>     corresponding dataloader.
> 
>     Args:
>         config (Config): An instance object of Config, used to record parameter information.
>         dataset (Dataset): An instance object of Dataset, which contains all interaction records.
>         save (bool, optional): If ``True``, it will call :func:`save_datasets` to save split dataset.
>             Defaults to ``False``.
> 
>     Returns:
>         tuple:
>             - train_data (AbstractDataLoader): The dataloader for training.
>             - valid_data (AbstractDataLoader): The dataloader for validation.
>             - test_data (AbstractDataLoader): The dataloader for testing.
>     """
>     model_type = config['MODEL_TYPE']
> 
>     es_str = [_.strip() for _ in config['eval_setting'].split(',')]
>     es = EvalSetting(config)
> 
>     kwargs = {}
>     if 'RS' in es_str[0]:
>         kwargs['ratios'] = config['split_ratio']
>         if kwargs['ratios'] is None:
>             raise ValueError('`ratios` should be set if `RS` is set')
>     if 'LS' in es_str[0]:
>         kwargs['leave_one_num'] = config['leave_one_num']
>         if kwargs['leave_one_num'] is None:
>             raise ValueError('`leave_one_num` should be set if `LS` is set')
>     kwargs['group_by_user'] = config['group_by_user']
>     getattr(es, es_str[0])(**kwargs)
> 
>     if es.split_args['strategy'] != 'loo' and model_type == ModelType.SEQUENTIAL:
>         raise ValueError('Sequential models require "loo" split strategy.')
> 
>     builded_datasets = dataset.build(es)
>     train_dataset, valid_dataset = builded_datasets
>     phases = ['train', 'test']
> 
>     if save:
>         save_datasets(config['checkpoint_dir'], name=phases, dataset=builded_datasets)
> 
>     kwargs = {}
>     if config['training_neg_sample_num']:
>         es.neg_sample_by(config['training_neg_sample_num'])
>         if model_type != ModelType.SEQUENTIAL:
>             sampler = Sampler(phases, builded_datasets, es.neg_sample_args['distribution'])
>         else:
>             sampler = RepeatableSampler(phases, dataset, es.neg_sample_args['distribution'])
>         kwargs['sampler'] = sampler.set_phase('train')
>         kwargs['neg_sample_args'] = copy.deepcopy(es.neg_sample_args)
>         if model_type == ModelType.KNOWLEDGE:
>             kg_sampler = KGSampler(dataset, es.neg_sample_args['distribution'])
>             kwargs['kg_sampler'] = kg_sampler
>     train_data = dataloader_construct(
>         name='train',
>         config=config,
>         eval_setting=es,
>         dataset=train_dataset,
>         dl_format=config['MODEL_INPUT_TYPE'],
>         batch_size=config['train_batch_size'],
>         shuffle=True,
>         **kwargs
>     )
> 
>     kwargs = {}
>     if len(es_str) > 1 and getattr(es, es_str[1], None):
>         getattr(es, es_str[1])()
>         if 'sampler' not in locals():
>             sampler = Sampler(phases, builded_datasets, es.neg_sample_args['distribution'])
>         kwargs['sampler'] = [sampler.set_phase('test')]
>         kwargs['neg_sample_args'] = copy.deepcopy(es.neg_sample_args)
>     valid_data = dataloader_construct(
>         name='evaluation',
>         config=config,
>         eval_setting=es,
>         dataset=valid_dataset,
>         batch_size=config['eval_batch_size'],
>         **kwargs
>     )
> 
>     return train_data, valid_data
> ```
> 
> Then you can use original split(5-folds) by setting `split_ratio` as [0.8, 0.2]

Thank you for replying so fast!

First I would like to make a correction on my former description, 
>  the spilt_ratio in evaluation setting would always take affect even though I set it as [0.8, 0.2, 0](which actually convert into [0.8, 0.1, 0.1]).

It turns out that I made a typo of `split_ratio` in the parameter_dict 😢 , then the default value of this parameter took affect. 

Back to my question, what I wanna do is use the MovieLens-100k datasets in original split which is already spilted into train_set and test_set, and I don't wanna change them by shuffling and re-sampling. I find that there are only two options which are Random Ordering and Temporal Ordering when it get to split the dataset.
The code you provided seems that just modify the func`data_preparation` to remove the valid_dataset part.

In that case, I guess the solution would be a customize a dataloader? Or modify the timestamp(or add a new column) in the dataset to make sure the data in the test_set get a bigger timestamp so that I can split them from the merged data, right?

**Comment by congx961** [comment]
Or can I use 2 different `Config` instance to create two diefferent `Dataset` and then convert it into `Dataloader` instance?
But I don't really know the work flow and necessary parameters, could you help me with that?

**Comment by chenyushuo** [comment]
@congx961 We just added ``config['benchmark_filename']`` to load pre-split dataset in #592. You just need to add 'benchmark_filename' to your configuration file (yaml), like this
```
benchmark_filename: ['part1', 'part2', 'part3']
```
And we will go into ``os.path.join(config['data_path'], config['dataset'])`` directory, and load three atomic file ``f'{config['dataset']}.part1.inter'``, ``f'{config['dataset']}.part2.inter'``, ``f'{config['dataset']}.part3.inter'``, which contains split user-item interaction, one by one. In this case, we will only apply normalize, remap-id, which will not delete the interaction in inter_feat.

Then, in ``data_preparation``, the dataset partition will keep up with the input ``config['benchmark_filename']``, that is to say ``train_data.dataset.inter_feat`` only contains ``f'{config['dataset']}.part1.inter'``, ``valid_data.dataset.inter_feat`` only contains ``f'{config['dataset']}.part2.inter'``,  ``test_data.dataset.inter_feat`` only contains ``f'{config['dataset']}.part3.inter'``.

By the way, the ``valid_data`` and ``test_data`` can have no interaction in their dataset, i.e. ``len(valid_data.dataset.inter_feat)`` and ``len(test_data.dataset.inter_feat)`` can be 0. If ``valid_data`` and ``test_data`` have no interaction in their dataset, we will not perform the corresponding valid or test procedure. Combined with the previous ``config['benchmark_filename']``, if pre-split dataset have no test data, your ``f'{config['dataset']}.part3.inter'`` can just have the first line which describe the feature info like other .inter file.

**Comment by congx961** [comment]
Thank you very much for your supporting!
It works!
我对你salute 🖖

**Comment by mayaKaplansky** [comment]
Thanks. Is part1 train, part2 validation, part3 test?

**Comment by chenyushuo** [comment]
@mayaKaplansky You are right. However, the sequential model does not support `config['benchmark_filename']` temporary.

**Comment by mayaKaplansky** [comment]
So what are my options when using sequential model for splitting?
Thanks!

**Comment by johnny12150** [comment]
> However, the sequential model does not support `config['benchmark_filename']` temporarily.

@chenyushuo 
Is there a timeline for adding pre-split dataset support to sequential models?
